### PR TITLE
Origin/lsp

### PR DESCRIPTION
**Pull request by lovestaco**
## What type of MR is this? (check all applicable)

- [x] 🍕 Feature
- [x] 🐛 Bug Fix
- [ ] 📝 Doc Update
- [ ] 🎨 Style
- [ ] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [ ] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI
- [ ] ⏩ Revert

## Description

* This MR updates the feature suggestenvs which was previously using `l2 -e='your_search_query'` using the LSP built for l2.
*  We will proceed to use LSP for future features.
```
		{
			"jsonrpc": "2.0",
			"id": 2,
			"method": "suggest/environmentVariables",
			"params": {
				"textDocument": {
					"uri": "file:///home/lovestaco/repos/Lama2/elfparser/ElfTestSuite/root_variable_override/api/y_0020_root_override.l2"
				},
				"position": {
					"line": 1,
					"character": 2
				},
				"searchQuery": "<your_search_query>"
			}
		}
```
* Relevant [Lama2 go binary release ](https://github.com/HexmosTech/Lama2/releases/tag/v1.5.8).

### Bug fix for envs within string enclosed env placeholder
image
 
### Unsupported error for windows
image

## Added tests?

- [ ] 👍 yes
- [x] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help

## Added to documentation?

- [ ] 📜 README.md
- [x] 🙅 no documentation needed

## Prettify the repositary

`npm run format`

- [x] 👍
- [ ] 🙅 no

### CODE REVIEW

File: src/extension.ts [hunk 1]
```diff
@@ -51,27 +65,48 @@ export function activate(context: vscode.ExtensionContext) {
     l2Version === undefined ||
     semver.lt(l2Version, MIN_VERSION_TO_CHECK)
   ) {
-    // The L2 version, if less than v1.5.0, does not support the `l2 -e <filepath>` feature.
-    // Therefore, we use the extension to fetch variables from the l2.env file for suggestions.
+    /**
+     * The L2 version, if less than v1.5.0, does not support the `l2 -e <filepath>` feature.
```
Comment (shrsv) [inline src/extension.ts]: Let's ditch this feature. We'll go 100% LSP from v1.5.9. We can remove it from backend in `v1.5.9` as well.

File: src/extension.ts [hunk 2]
```diff
@@ -51,27 +65,48 @@ export function activate(context: vscode.ExtensionContext) {
...
+     * Therefore, we use the extension to fetch variables from the l2.env file for suggestions.
+     */
     getDotENVS();
     suggestEnvVariables = suggestENVSFromDotEnv();
   } else {
-    // L2 versions greater than or equal to v1.5.0 support the `l2 -e <filepath>` feature.
-    // Therefore, we fetch variables from both the l2.env and l2config.env files for suggestions using this command.
-    suggestEnvVariables = lama2RegisterCompletionItemProvider();
+    /**
+     * L2 versions greater than or equal to v1.5.0 support the `l2 -e <filepath>` feature.
```
Comment (shrsv) [inline src/extension.ts]: No need to retain this as well. Only LSP going forward.

File: src/lsp/methods/lspLifecycles.ts [hunk 3]
```diff
@@ -0,0 +1,51 @@
+import * as vscode from "vscode";
+import { ChildProcess } from "child_process";
+import {
+  ILSPRequestDetails,
+  sendRequestToLSPReadResponse,
+} from "../request/generalRequest";
+
+export function initlizeServer(langServer: ChildProcess, requestId: number) {
```
Comment (shrsv) [inline src/lsp/methods/lspLifecycles.ts]: Spelling.

File: src/lsp/methods/suggestEnvironmentVars.ts [hunk 4]
```diff
@@ -1,8 +1,14 @@
...
+import { ChildProcess, execSync } from "child_process";
 import * as path from "path";
 import * as fs from "fs";
-import triggers from "./triggers";
+import triggers from "../../triggers";
+import {
+  ErrorCodes,
+  IJSONRPCResponse,
+  logToChannel,
+} from "../response/generalResponse";
```
Comment (shrsv) [inline src/lsp/methods/suggestEnvironmentVars.ts:10]: We can get imports based on project root everywhere if possible.

File: src/lsp/methods/suggestEnvironmentVars.ts [hunk 5]
```diff
@@ -94,41 +70,136 @@ function createSuggestionsArray(
...
+
+export function triggerSuggestionInTemplateLiteral() {
+  return vscode.workspace.onDidChangeTextDocument((event) => {
+    const editor = vscode.window.activeTextEditor;
+    if (editor && event.document === editor.document) {
+      const cursorPosition = editor.selection.active;
+      const lineText = editor.document.lineAt(cursorPosition.line).text;
 
-      return completions;
+      if (isCursorInsideTemplateLiteral(lineText, cursorPosition.character)) {
```
Comment (shrsv) [inline src/lsp/methods/suggestEnvironmentVars.ts]: `isCursorInVarContainer`

File: src/lsp/request/generalRequest.ts [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+export interface IJSONRPCRequest {
+  jsonrpc: "2.0";
+  id: number;
+  method: IJSONRPCMethod;
+  params: any;
+}
+
+// Function to send requests to the Language Server Protocol and read responses
+export function sendRequestToLSPReadResponse(
```
Comment (shrsv) [inline src/lsp/request/generalRequest.ts]: Simplify name - maybe something like: `askLSP` (and response type is clear by return type)

File: src/lsp/request/generalRequest.ts [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+      params,
+    };
+
+    logToChannel({ msg: "Sending request: ", dataObject: request });
+    const requestString = JSON.stringify(request);
+    process.stdin.write(requestString + "\n");
+
+    const chunks: any[] = [];
+    process.stdout.on("data", (data) => {
+      chunks.push(data);
```
Comment (shrsv) [inline src/lsp/request/generalRequest.ts]: You can put these event handlers as separate as functions; as of now this function is a bit too long and not super easy to read.

File: src/lsp/response/generalResponse.ts [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+  result?: any;
+  error?: any;
+}
+
+// Helper function to log messages to the output channel
+export function logToChannel(options: ILogOptions) {
+  const { msg: message, logType = "info", dataObject, dataString } = options;
+
+  const timestamp = new Date().toLocaleTimeString();
+  const logTypeInfo = LOG_TYPES[logType.toLowerCase()] || LOG_TYPES["info"];
```
Comment (shrsv) [inline src/lsp/response/generalResponse.ts:48]: Isn't L45 setting `logType` default value? I don't think we need `||`

Comment (lovestaco) [reply src/lsp/response/generalResponse.ts:48]: Acts like default value 
```
   logToChannel({
        msg: "Received envs from server",
        dataObject: response,
      });
 ```

File: src/lsp/request/generalRequest.ts (hunk 7 above)
Comment (lovestaco) [reply src/lsp/request/generalRequest.ts]: Right addressed it

File: src/lsp/request/generalRequest.ts (hunk 6 above)
Comment (lovestaco) [reply src/lsp/request/generalRequest.ts]: okay

File: src/lsp/methods/suggestEnvironmentVars.ts (hunk 5 above)
Comment (lovestaco) [reply src/lsp/methods/suggestEnvironmentVars.ts]: done

File: src/extension.ts (hunk 2 above)
Comment (lovestaco) [reply src/extension.ts]: Cool

File: src/extension.ts (hunk 1 above)
Comment (lovestaco) [reply src/extension.ts]: Done

File: src/lsp/methods/suggestEnvironmentVars.ts (hunk 4 above)
Comment (lovestaco) [reply src/lsp/methods/suggestEnvironmentVars.ts:10]: Will do in the next MR, tried, taking a lot of time

**Pull request merged**